File: java-samples/Advanced/snakesandladders-cordapp/oracle-flows/src/main/java/net/corda/samples/snl/oracle/flows/OracleSignatureFlow.java
package net.corda.samples.snl.oracle.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.crypto.TransactionSignature;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.flows.FlowSession;
import net.corda.core.flows.InitiatingFlow;
import net.corda.core.identity.Party;
import net.corda.core.transactions.FilteredTransaction;

@InitiatingFlow
public class OracleSignatureFlow extends FlowLogic<TransactionSignature> {

    private final Party oracle;
    private final FilteredTransaction ftx;

    public OracleSignatureFlow(Party oracle, FilteredTransaction ftx) {
        this.oracle = oracle;
        this.ftx = ftx;
    }

    @Suspendable
    @Override
    public TransactionSignature call() throws FlowException {
        FlowSession session = initiateFlow(oracle);
        return session.sendAndReceive(TransactionSignature.class, ftx).unwrap(it -> it);
    }
}

File: kotlin-samples/Advanced/snakesandladders-cordapp/oracle-flows/src/main/kotlin/net/corda/samples/snl/oracle/flows/OracleSignatureFlow.kt
package net.corda.samples.snl.oracle.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.crypto.TransactionSignature
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.transactions.FilteredTransaction
import net.corda.core.utilities.unwrap
import net.corda.samples.snl.diceservice.DiceRollService


@InitiatingFlow
class OracleSignatureFlow(private val oracle: Party, private val ftx: FilteredTransaction) : FlowLogic<TransactionSignature?>() {
    @Suspendable
    @Throws(FlowException::class)
    override fun call() = initiateFlow(oracle).sendAndReceive(TransactionSignature::class.java, ftx).unwrap { it }

}

@InitiatedBy(OracleSignatureFlow::class)
class OracleSignatureFlowHandler(private val requestSession: FlowSession) : FlowLogic<Unit>() {
    @Suspendable
    @Throws(FlowException::class)
    override fun call(): Unit {
        val transaction = requestSession.receive(FilteredTransaction::class.java).unwrap{ it-> it }
        var signature: TransactionSignature? = null
        signature = try {
            serviceHub.cordaService(DiceRollService::class.java).sign(transaction)
        } catch (e: Exception) {
            throw FlowException(e)
        }
        requestSession.send(signature)
        return
    }
}
